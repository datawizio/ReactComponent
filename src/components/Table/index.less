@import "~antd/es/table/style/index";
@import "~antd/es/empty/style/index";
@import "~antd/es/pagination/style/index";

@import "../Select/index";
@import "../Loader/index";
@import "../../styles/variables";

.dw-table {
  &__wrapper {
    overflow: auto;
  }

  &__toolbar {
    padding: 0 5px;
    display: flow-root;
    position: relative;

    .table-toolbar--left {
      float: left;
      margin-right: 3px;
    }

    .table-toolbar--right {
      float: right;
      margin-right: 3px;
    }
  }

  .dw-table__column {
    width: 300px;
    transition: none !important;

    top: 0;
    z-index: 5;
    overflow: hidden;
    position: sticky !important;

    &--resizable {
      resize: horizontal;
    }

    &--drop-hover {
      background-color: tint(@primary-color, 80%) !important;
    }

    &--fixed {
      z-index: 10;

      &--left {
        left: 0;
      }

      &--right {
        right: 0;
      }
    }

    &-content {
      display: flex;
      flex-direction: row;
      align-items: center;
    }

    &-sort-sequence {
      color: @primary-color;
    }

    .ant-table-filter-trigger-container {
      background: transparent !important;
      right: 10px;
    }

    .ant-table-filter-column-title {
      padding: 8px !important;
    }
  }

  &__cell-data {
    display: flex;
    align-items: center;
  }

  table {
    table-layout: fixed;
    width: max-content !important;
  }

  td {
    .row-prefix {
      float: left;
      margin-right: 8px;
    }

    .ant-table-row-expand-icon {
      float: none;
    }

    position: relative;

    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  &--loading {
    table {
      .ant-empty {
        display: none;
      }
    }
  }

  &--empty {
    table {
      width: 100% !important;
      height: 100% !important;

      thead {
        display: none;
      }

      td {
        left: 50%;
        border: none !important;
        position: absolute !important;
      }
    }
  }

  .ant-table-bordered {
    .ant-table-cell {
      border-right: 1px solid #f0f0f0;
    }
  }

  .ant-table {
    .set-position-header(10, 55);

    &-small {
      .set-position-header(10, 39);
    }

    &-middle {
      .set-position-header(10, 47);
    }
  }
}

.set-columns-positions(@n, @size, @i: 2) when (@i =< @n) {
  tr:nth-child(@{i}) {
    .dw-table__column {
      top: unit(@size * (@i - 1), px);
    }
  }

  .set-columns-positions(@n, @size, (@i + 1));
}

.set-position-header(@n, @size) {
  thead {
    .set-columns-positions(@n, @size);
  }
}
