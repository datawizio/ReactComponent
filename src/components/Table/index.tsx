import clsx from "clsx";
import * as React from "react";
import { Table as AntdTable } from "antd";
import { reducer, initializer } from "./reducer";

import useColumns from "./hooks/useColumns";
import useDataSource from "./hooks/useDataSource";
import usePropsToState from "./hooks/usePropsToState";

import {
  useMemo,
  useReducer,
  useContext,
  useCallback,
  useImperativeHandle
} from "react";

import { TableContext } from "./context";
import ConfigContext from "../ConfigProvider/context";

import Loader from "../Loader";
import Cell from "./components/Cell";
import Column from "./components/Column";
import ToolBar from "./components/ToolBar";
import TableWrapper from "./components/TableWrapper";
import {
  LoadingOutlined,
  RightOutlined,
  DownOutlined
} from "@ant-design/icons";

import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";

import {
  basicSortHandler,
  basicFilterHandler,
  basicSearchHandler
} from "./utils/handlers";

import { TableProps, FCTable, TableRef } from "./types";

import "./index.less";
import useAsyncProviders from "./hooks/useAsyncProviders";

const Table = React.forwardRef<TableRef, TableProps>((props, ref) => {
  const {
    style,
    width,
    height,
    locale,
    isNested,
    showExpandIcon,
    children,
    components,
    dataProvider,
    showSizeChanger,
    dataProviderDeps,
    templatesProvider,
    rowChildrenProvider,
    nestedTableProvider,
    ...restProps
  } = props;

  const { translate } = useContext(ConfigContext);

  const [baseState, dispatch] = useReducer(reducer, props, initializer);

  const columnsState = useColumns(baseState, props);

  const dataSourceState = useDataSource(baseState, props);

  const state = {
    ...baseState,
    ...dataSourceState,
    ...columnsState,
    loading: false
  };

  const fetchData = useAsyncProviders(state, dispatch, props);

  usePropsToState(dispatch, props);

  const handleChangeTable = useCallback<TableProps["onChange"]>(
    (pagination, filters, sorter) => {
      dispatch({ type: "filter", payload: filters });
      dispatch({ type: "sort", payload: [].concat(sorter) });
      dispatch({ type: "paginate", payload: pagination as any });
    },
    []
  );

  const handleExpandRow = useCallback<TableProps["onExpand"]>(
    async (isExpanded, row) => {
      let toogle = true;
      if ((isNested && isNested(row)) || (!isNested && nestedTableProvider)) {
        if (!row.nested) {
          dispatch({
            type: "addLoadingRow",
            payload: row.key
          });
          const result = await nestedTableProvider(row);
          if (!result) toogle = false;
          dispatch({
            type: "setNestedTable",
            payload: [row, result]
          });
        }
      } else if (rowChildrenProvider && row.children && !row.children.length) {
        dispatch({
          type: "addLoadingRow",
          payload: row.key
        });
        const children = await rowChildrenProvider(row);
        dispatch({
          type: "setRowChildren",
          payload: [row, children.length ? children : undefined]
        });
      }

      toogle &&
        dispatch({
          type: isExpanded ? "expandRow" : "collapseRow",
          payload: row
        });
    },
    [rowChildrenProvider, nestedTableProvider, isNested]
  );

  const totalRenderer = useCallback(
    (total, [current, to]) => {
      return translate(locale.total, { current, to, total });
    },
    [translate, locale.total]
  );

  const expandIconRender = useCallback(
    ({ expanded, onExpand, record, prefixCls, expandable }) => {
      const iconPrefix = `${prefixCls}-row-expand-icon`;
      if (state.loadingRows[record.key]) return <LoadingOutlined />;
      let icon = null;
      let internalExpandable =
        expandable || (showExpandIcon && showExpandIcon(record));
      if (internalExpandable) {
        icon = <RightOutlined />;
        if (expanded) {
          icon = <DownOutlined />;
        }
      }

      return (
        <button
          type="button"
          onClick={e => {
            onExpand(record, e!);
            e.stopPropagation();
          }}
          className={clsx(iconPrefix, {
            [`${iconPrefix}-spaced`]: !internalExpandable,
            [`${iconPrefix}-expanded`]: expandable && expanded,
            [`${iconPrefix}-collapsed`]: expandable && !expanded
          })}
          aria-label={expanded ? locale.collapse : locale.expand}
        >
          {icon}
        </button>
      );
    },
    [state.loadingRows]
  );

  const customComponents = useMemo<TableProps["components"]>(
    () => ({
      ...components,
      table: props => <TableWrapper {...props} style={{ height, width }} />,
      header: {
        cell: props =>
          Boolean(props.model) ? <Column {...props} /> : <th {...props} />
      },
      body: {
        cell: props => <Cell {...props} />
      }
    }),
    [height, width, components]
  );

  const className = useMemo(
    () =>
      clsx(
        "dw-table",
        {
          "dw-table--loading": baseState.loading,
          "dw-table--empty": !state.dataSource.length,
          "dw-table--nestedable": props.expandable?.expandedRowRender
        },
        props.className
      ),
    [baseState.loading, props.className, state.dataSource.length]
  );

  useImperativeHandle(ref, () => ({
    async reload() {
      dispatch({ type: "loading", payload: true });
      await fetchData();
      dispatch({ type: "loading", payload: false });
    },
    updateRow(rowKey: string, data: any) {
      dispatch({ type: "updateRow", payload: [rowKey, data] });
    },
    addLoadingRow(rowKey: string) {
      dispatch({ type: "addLoadingRow", payload: rowKey });
    },
    resetPagination() {
      dispatch({ type: "resetPagination" });
    }
  }));

  return (
    <DndProvider backend={HTML5Backend}>
      <TableContext.Provider
        value={{
          tableProps: props,
          tableState: state,
          dispatch: dispatch,
          baseTableState: baseState
        }}
      >
        <Loader loading={Boolean(baseState.loading)}>
          {children}
          <AntdTable
            {...restProps}
            {...state}
            expandIcon={expandIconRender}
            className={className}
            onExpand={handleExpandRow}
            onChange={handleChangeTable}
            components={customComponents}
            pagination={
              props.pagination === false
                ? false
                : {
                    ...state.pagination,
                    ...props.pagination,
                    showTotal: totalRenderer
                  }
            }
          />
        </Loader>
      </TableContext.Provider>
    </DndProvider>
  );
}) as FCTable;

Table.defaultProps = {
  bordered: true,
  sortable: true,
  isResizableColumns: true,

  showSorterTooltip: false,

  className: "",
  size: "small",
  width: "auto",
  height: "auto",
  tableLayout: "fixed",

  locale: {
    total: "TABLE_TOTAL"
  },

  dTypesConfig: {},
  columnsConfig: {},
  cellRenderProps: {},

  columns: [],
  dataSource: [],
  pageSizeOptions: ["20", "35", "50", "100"],

  showSizeChanger: true,
  sortHandler: basicSortHandler,
  filterHandler: basicFilterHandler,
  searchHandler: basicSearchHandler
};

Table.ToolBar = ToolBar;

export default Table;
