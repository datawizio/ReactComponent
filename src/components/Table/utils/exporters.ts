import { TableState, ISheet } from "../types";
import ExcelJS from "exceljs/dist/exceljs.min.js";
import { defineCellType } from "./utils";

function getDeepMaxLevel(columns) {
  let maxLevel = 1;

  (function rec(columns, level = 1) {
    columns.forEach(column => {
      if (column.children && column.children.length)
        rec(column.children, level + 1);
      if (level > maxLevel) maxLevel = level;
    });
  })(columns);

  return maxLevel;
}

function getDeepChildren(columns) {
  let deepChildren = [];

  (function rec(columns) {
    columns.forEach(column => {
      if (column.children && column.children.length) rec(column.children);
      else deepChildren.push(column);
    });
  })(columns);

  return deepChildren;
}

function aroundBorderCell(cell) {
  cell.border = {
    top: { style: "thin" },
    left: { style: "thin" },
    bottom: { style: "thin" },
    right: { style: "thin" }
  };
}

function alignCenterCellText(cell) {
  cell.alignment = {
    horizontal: "center",
    vertical: "middle"
  };
}

function fillBackgroundCell(cell, color) {
  cell.fill = {
    type: "pattern",
    pattern: "solid",
    fgColor: {
      argb: color
    },
    bgColor: {
      argb: color
    }
  };
}

export async function exportTableToXLSX(
  tableState: TableState | null,
  filename: string,
  sheetName?: string,
  cellRenderProps?: string,
  additionalSheet?: ISheet
): Promise<BlobPart> {
  const { columns, columnsMap, dataSource, dTypesConfig } = tableState;

  const formattedSheetName = (sheetName || filename)
    .replace(/[`~!@#$%^&*()|+\-=?;:'",.<>\{\}\[\]\\\/]/gi, "")
    .replace(".xlsx", "");
  const wb = new ExcelJS.Workbook(); // make a workbook
  const ws = wb.addWorksheet(formattedSheetName); // make a worksheet

  (function addAdditionalSheet(sheet) {
    if (!sheet?.name || !sheet?.state) return;
    const newSheet = wb.addWorksheet(sheet.name);
    const columns = Object.keys(sheet.state);

    newSheet.columns = columns.map(columnName => {
      return { header: columnName, key: columnName, width: 20 };
    });

    columns.forEach(columnKey => {
      const col = newSheet.getColumn(columnKey);
      col.values = [columnKey, ...sheet.state[columnKey]];
    });

    newSheet.eachRow(Row => {
      Row.eachCell(Cell => {
        Cell.alignment = { wrapText: true };
      });
    });

    newSheet.getRow(1).eachCell(cell => {
      aroundBorderCell(cell);
      alignCenterCellText(cell);
      fillBackgroundCell(cell, "efefef");
    });
  })(additionalSheet);

  const columnsMaxLevel = getDeepMaxLevel(columns);
  const defaultValues = {};
  // recursive draw columns
  (function drawColumns(columns, rowIdx = 1, colIdx = 1) {
    columns.forEach(column => {
      const row = ws.getRow(rowIdx);
      const cell = row.getCell(colIdx);

      // set column title
      cell.value = column.title;

      aroundBorderCell(cell);
      alignCenterCellText(cell);
      fillBackgroundCell(cell, "efefef");

      const dType =
        dTypesConfig[defineCellType({}, columnsMap[column.dataIndex])];
      if (dType && typeof dType.defaultValue !== "undefined") {
        defaultValues[column.dataIndex] = 0;
      }

      // check children for correct merge cells width
      if (column.children && column.children.length) {
        const childrenCount = getDeepChildren(column.children).length;
        const nextCellIdx = colIdx + (childrenCount - 1);
        const nextCell = row.getCell(nextCellIdx);

        ws.mergeCells(`${cell.address}:${nextCell.address}`);
        drawColumns(column.children, rowIdx + 1, colIdx);

        colIdx = colIdx + childrenCount;
      } else {
        // if not child stretch height of cell to max children level of columns
        const letter = cell._column.letter;
        ws.mergeCells(`${cell.address}:${letter}${columnsMaxLevel}`);
        ++colIdx;
      }
    });
  })(columns);
  // add columns
  ws.columns = getDeepChildren(columns).map(({ dataIndex, width }) => ({
    key: dataIndex,
    width: Number(width || 200) / 10
  }));
  // recursive add row for tree table
  (function drawRows(rows, level = 0) {
    rows.forEach(rowData => {
      const { children, key, ...restRowData } = rowData;
      const margin = "  ".repeat(level);
      const firstColumn = ws.getColumn(1).key;

      // to string by dType
      const rowToDraw = Object.entries(restRowData).reduce(
        (acc, [columnKey, cellValue]) => {
          const dType =
            dTypesConfig[defineCellType(cellValue, columnsMap[columnKey])];
          acc[columnKey] = "unknown";
          if (dType) {
            acc[columnKey] = dType.toExcel
              ? dType.toExcel(cellValue, rowData, columnKey, cellRenderProps)
              : dType.toString(cellValue);
          }
          return acc;
        },
        {}
      );
      const row = ws.addRow({
        ...defaultValues,
        ...rowToDraw,
        key,
        [firstColumn]: margin + rowToDraw[firstColumn]
      });

      row.hidden = level !== 0;
      row.outlineLevel = level;

      row.eachCell(cell => {
        aroundBorderCell(cell);
      });

      if (children) drawRows(children, level + 1);
    });
  })(dataSource);

  return await wb.xlsx.writeBuffer();
}
