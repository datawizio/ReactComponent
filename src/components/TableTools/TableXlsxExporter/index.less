@import "../../Button/index";
@import "../../../styles/variables";

.table-xlsx-exporter {
  &__icon {
    color: @primary-color;
  }
}
