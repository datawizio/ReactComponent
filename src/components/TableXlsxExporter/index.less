@import "../Button/index";
@import "~antd/es/message/style/index";
@import "../../styles/variables";

.table-xlsx-exporter {
  &__icon {
    color: @primary-color;
  }
}
