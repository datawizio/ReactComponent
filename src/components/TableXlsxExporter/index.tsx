import * as React from "react";
import { message } from "antd";
import { saveAs } from "file-saver";
import { TableState } from "../Table/types";

import { useContext, useCallback } from "react";
import { TableContext } from "../Table/context";
import ConfigContext from "../ConfigProvider/context";

import Button, { ButtonProps } from "../Button";
import { DownloadOutlined } from "@ant-design/icons";

import "./index.less";

export interface TableXlsxExporterProps extends ButtonProps {
  filename?: string;
  exportHandler?: (
    tableState: TableState,
    filename: string
  ) => Promise<BlobPart> | Promise<void>;
  exportHandlerCallback?: (
    fileData: any,
    filename: string
  ) => Promise<void>;
}

const TableXlsxExporter: React.FC<TableXlsxExporterProps> = props => {
  const { exportHandler, exportHandlerCallback, filename, ...restProps } = props;

  const { translate } = useContext(ConfigContext);
  const { tableState } = useContext(TableContext);

  const handleExport = useCallback(async () => {
    if (exportHandler) {
      const messageKey = "exporting-" + filename;

      message.loading({ content: translate("LOADING"), key: messageKey });

      const fileData = await exportHandler(tableState, filename);
      fileData && saveAs(new Blob([fileData]), filename);

      message.success({ content: translate("SUCCESS"), key: messageKey });

      if (exportHandlerCallback && fileData) {
        await exportHandlerCallback(fileData, filename);
      }
    }
  }, [tableState, translate, exportHandler, filename, exportHandlerCallback]);

  return (
    <div className="table-xlsx-exporter table-toolbar--right">
      <Button
        {...restProps}
        border={false}
        onClick={handleExport}
        title={translate("SAVE_BTN_TITLE")}
      >
        <DownloadOutlined className={"table-xlsx-exporter__icon"} />
      </Button>
    </div>
  );
};

TableXlsxExporter.defaultProps = {
  filename: "exported_table.xlsx"
};

export default TableXlsxExporter;
