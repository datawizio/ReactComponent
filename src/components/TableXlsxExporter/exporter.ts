import { saveAs } from "file-saver";
import { TableState } from "../Table/types";
import ExcelJS from "exceljs/dist/exceljs.min.js";
import { defineCellType } from "../Table/utils/utils";

function getDeepMaxLevel(columns) {
  let maxLevel = 1;

  (function rec(columns, level = 1) {
    columns.forEach(column => {
      if (column.children && column.children.length)
        rec(column.children, level + 1);
      if (level > maxLevel) maxLevel = level;
    });
  })(columns);

  return maxLevel;
}

function getDeepChildren(columns) {
  let deepChildren = [];

  (function rec(columns) {
    columns.forEach(column => {
      if (column.children && column.children.length) rec(column.children);
      else deepChildren.push(column);
    });
  })(columns);

  return deepChildren;
}

function aroundBorderCell(cell) {
  cell.border = {
    top: { style: "thin" },
    left: { style: "thin" },
    bottom: { style: "thin" },
    right: { style: "thin" }
  };
}

function alignCenterCellText(cell) {
  cell.alignment = {
    horizontal: "center",
    vertical: "middle"
  };
}

function fillBackgroundCell(cell, color) {
  cell.fill = {
    type: "pattern",
    pattern: "solid",
    fgColor: {
      argb: color
    },
    bgColor: {
      argb: color
    }
  };
}

export async function exportTableToXLSX(
  columns: TableState["columns"],
  dataSource: TableState["dataSource"],
  columnsMap: TableState["columnsMap"],
  dTypesConfig: TableState["dTypesConfig"],
  filename: string,
  sheetName?: string
) {
  const wb = new ExcelJS.Workbook(); // make a workbook
  const ws = wb.addWorksheet(sheetName || filename); // make a worksheet
  const columnsMaxLevel = getDeepMaxLevel(columns);

  // recursive draw columns
  (function drawColumns(columns, rowIdx = 1, colIdx = 1) {
    columns.forEach(column => {
      const row = ws.getRow(rowIdx);
      const cell = row.getCell(colIdx);

      // set column title
      cell.value = column.title;

      aroundBorderCell(cell);
      alignCenterCellText(cell);
      fillBackgroundCell(cell, "efefef");

      // check children for correct merge cells width
      if (column.children && column.children.length) {
        const childrenCount = getDeepChildren(column.children).length;
        const nextCellIdx = colIdx + (childrenCount - 1);
        const nextCell = row.getCell(nextCellIdx);

        ws.mergeCells(`${cell.address}:${nextCell.address}`);
        drawColumns(column.children, rowIdx + 1, colIdx);

        colIdx = colIdx + childrenCount;
      } else {
        // if not child stretch height of cell to max children level of columns
        const letter = cell._column.letter;
        ws.mergeCells(`${cell.address}:${letter}${columnsMaxLevel}`);
        ++colIdx;
      }
    });
  })(columns);

  // add columns
  ws.columns = getDeepChildren(columns).map(({ dataIndex, width }) => ({
    key: dataIndex,
    width: Number(width || 200) / 10
  }));

  // recursive add row for tree table
  (function drawRows(rows, level = 0) {
    rows.forEach(rowData => {
      const { children, key, ...restRowData } = rowData;
      const margin = "  ".repeat(level);
      const firstColumn = ws.getColumn(1).key;

      // to string by dType
      const rowToDraw = Object.entries(restRowData).reduce(
        (acc, [columnKey, cellValue]) => {
          const dType =
            dTypesConfig[defineCellType(cellValue, columnsMap[columnKey])];
          acc[columnKey] = dType ? dType.toString(cellValue) : "unknown";
          return acc;
        },
        {}
      );

      const row = ws.addRow({
        ...rowToDraw,
        key,
        [firstColumn]: margin + rowToDraw[firstColumn]
      });

      row.hidden = level !== 0;
      row.outlineLevel = level;

      row.eachCell(cell => {
        aroundBorderCell(cell);
      });

      if (children) drawRows(children, level + 1);
    });
  })(dataSource);

  saveAs(new Blob([await wb.xlsx.writeBuffer()]), filename + ".xlsx");
}
