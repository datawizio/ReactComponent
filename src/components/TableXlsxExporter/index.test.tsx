import "jsdom-global/register";
import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { render, fireEvent } from "@testing-library/react";
import TableXlsxExporter from "./index";
import Table from "../Table";
import {
  getStaticColumn,
  getStaticDataSource,
  sortable
} from "../Table/__mocks__";
import { exportTableToXLSX } from "./exporters";

const mockTableProps = {
  height: "auto",
  width: "auto",
  dTypesConfig: {},

  dataProvider: jest.fn(() => {
    return {
      columns: [
        {
          title: "Name",
          key: "name",
          dataIndex: "name"
        }
      ],
      dataSource: [
        {
          key: "1",
          name: "John"
        }
      ]
    };
  })
};

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
});

const setUp = (props?) => mount(<TableXlsxExporter {...props} />);

describe("TableXlsxExporter component", () => {
  let component;

  beforeEach(() => {
    component = setUp();
  });

  it("should render corectly", () => {
    expect(component).toMatchSnapshot();
  });

  it("should render in Table ToolBar corectly", () => {
    const { container } = render(
      //@ts-ignore
      <Table {...mockTableProps}>
        <Table.ToolBar>
          <TableXlsxExporter exportHandler={exportTableToXLSX} />
        </Table.ToolBar>
      </Table>
    );
    expect(container).toMatchSnapshot();
  });

  it("should export corectly", async () => {
    const mockProps = {
      height: "auto",
      width: "auto",

      sortable,
      columns: getStaticColumn(),
      dataSource: getStaticDataSource()
    };

    const wrapper = render(
      //@ts-ignore
      <Table {...mockProps}>
        <Table.ToolBar>
          <TableXlsxExporter exportHandler={exportTableToXLSX} />
        </Table.ToolBar>
      </Table>
    );
    const saveButton = wrapper.getByTitle("SAVE_BTN_TITLE");
    act(() => {
      fireEvent.click(saveButton);
    });
    expect(
      document.querySelectorAll(".ant-message-notice")[0].textContent
    ).toBe("LOADING");
  });
});
