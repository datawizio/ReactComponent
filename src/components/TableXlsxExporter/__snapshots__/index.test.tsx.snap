// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableXlsxExporter component should render corectly 1`] = `
<TableXlsxExporter
  filename="exported_table.xlsx"
>
  <div
    className="table-xlsx-exporter table-toolbar--right"
  >
    <Button
      border={false}
      icon={
        <ForwardRef(DownloadOutlined)
          className="table-xlsx-exporter__icon"
        />
      }
      loading={false}
      onClick={[Function]}
      title="SAVE_BTN_TITLE"
    >
      <Button
        className="dw-btn no-border"
        icon={
          <ForwardRef(DownloadOutlined)
            className="table-xlsx-exporter__icon"
          />
        }
        loading={false}
        onClick={[Function]}
        title="SAVE_BTN_TITLE"
      >
        <Wave>
          <button
            className="ant-btn ant-btn-icon-only dw-btn no-border"
            onClick={[Function]}
            title="SAVE_BTN_TITLE"
            type="button"
          >
            <ForwardRef(DownloadOutlined)
              className="table-xlsx-exporter__icon"
            >
              <AntdIcon
                className="table-xlsx-exporter__icon"
                icon={
                  Object {
                    "icon": Object {
                      "attrs": Object {
                        "focusable": "false",
                        "viewBox": "64 64 896 896",
                      },
                      "children": Array [
                        Object {
                          "attrs": Object {
                            "d": "M505.7 661a8 8 0 0012.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                          },
                          "tag": "path",
                        },
                      ],
                      "tag": "svg",
                    },
                    "name": "download",
                    "theme": "outlined",
                  }
                }
              >
                <span
                  aria-label="download"
                  className="anticon anticon-download table-xlsx-exporter__icon"
                  role="img"
                >
                  <IconReact
                    icon={
                      Object {
                        "icon": Object {
                          "attrs": Object {
                            "focusable": "false",
                            "viewBox": "64 64 896 896",
                          },
                          "children": Array [
                            Object {
                              "attrs": Object {
                                "d": "M505.7 661a8 8 0 0012.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z",
                              },
                              "tag": "path",
                            },
                          ],
                          "tag": "svg",
                        },
                        "name": "download",
                        "theme": "outlined",
                      }
                    }
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="download"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      key="svg-download"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M505.7 661a8 8 0 0012.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                        key="svg-download-svg-0"
                      />
                    </svg>
                  </IconReact>
                </span>
              </AntdIcon>
            </ForwardRef(DownloadOutlined)>
          </button>
        </Wave>
      </Button>
    </Button>
  </div>
</TableXlsxExporter>
`;

exports[`TableXlsxExporter component should render in Table ToolBar corectly 1`] = `
<div>
  <div
    class="dw-table-container"
  >
    <div
      class="ant-spin-nested-loading"
    >
      <div>
        <div
          class="ant-spin ant-spin-spinning datawiz-loader"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin ant-spin-dot"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        class="ant-spin-container ant-spin-blur"
      >
        <div
          class="dw-table__toolbar"
        >
          <div
            class="table-xlsx-exporter table-toolbar--right"
          >
            <button
              class="ant-btn ant-btn-icon-only dw-btn no-border"
              title="SAVE_BTN_TITLE"
              type="button"
            >
              <span
                aria-label="download"
                class="anticon anticon-download table-xlsx-exporter__icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="download"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M505.7 661a8 8 0 0012.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
        <div
          class="ant-table-wrapper dw-table dw-table--loading dw-table--empty"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <div
                class="ant-table ant-table-small ant-table-bordered ant-table-empty ant-table-layout-fixed"
              >
                <div
                  class="ant-table-container"
                >
                  <div
                    class="ant-table-content"
                  >
                    <div
                      class="dw-table__wrapper"
                      style="width: auto; height: auto;"
                    >
                      <table>
                        <colgroup />
                        <thead
                          class="ant-table-thead"
                        >
                          <tr>
                            <th
                              class="ant-table-cell"
                            />
                          </tr>
                        </thead>
                        <tbody
                          class="ant-table-tbody"
                        >
                          <tr
                            class="ant-table-placeholder"
                          >
                            <td
                              class="ant-table-cell"
                            >
                              <div
                                class="dw-table__cell"
                              >
                                <div
                                  class="ant-empty ant-empty-normal"
                                >
                                  <div
                                    class="ant-empty-image"
                                  >
                                    <svg
                                      class="ant-empty-img-simple"
                                      height="41"
                                      viewBox="0 0 64 41"
                                      width="64"
                                      xmlns="http://www.w3.org/2000/svg"
                                    >
                                      <g
                                        fill="none"
                                        fill-rule="evenodd"
                                        transform="translate(0 1)"
                                      >
                                        <ellipse
                                          class="ant-empty-img-simple-ellipse"
                                          cx="32"
                                          cy="33"
                                          rx="32"
                                          ry="7"
                                        />
                                        <g
                                          class="ant-empty-img-simple-g"
                                          fill-rule="nonzero"
                                        >
                                          <path
                                            d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                          />
                                          <path
                                            class="ant-empty-img-simple-path"
                                            d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                          />
                                        </g>
                                      </g>
                                    </svg>
                                  </div>
                                  <div
                                    class="ant-empty-description"
                                  >
                                    No Data
                                  </div>
                                </div>
                              </div>
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
